firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/CustomModelDownloadService.java

```diff
@@ -0,0 +1,296 @@
+            connection.setRequestProperty(API_KEY_HEADER, apiKey);
+
+            return fetchDownloadDetails(modelName, connection);
+          });
+
+    } catch (Exception e) {
+      // TODO(annz) update to better error handling (use FirebaseMLExceptions)
+      throw new Exception("Error reading custom model from download service: " + e.getMessage(), e);
+    }
+  }
+
+  @VisibleForTesting
+  static long parseTokenExpirationTimestamp(String expiresIn) {
+    if (expiresIn == null || expiresIn.length() == 0) {
+      return 0;
+    }
+
+    try {
+      String isoDatePattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+      SimpleDateFormat iso8601Format = new SimpleDateFormat(isoDatePattern, Locale.US);
+      iso8601Format.setTimeZone(TimeZone.getTimeZone("UTC"));
+      Date date = iso8601Format.parse(expiresIn);
+      return date.getTime();
+    } catch (ParseException pe) {
+      // log error and maybe throw an error
+      Log.w(TAG, "unable to parse datetime:" + expiresIn, pe);
+      return 0;
+    }
+  }
+
+  private Task<CustomModel> fetchDownloadDetails(String modelName, HttpURLConnection connection)
+      throws Exception {
+
+    connection.connect();
+    int httpResponseCode = connection.getResponseCode();
+
+    if ((httpResponseCode != HttpURLConnection.HTTP_OK)
+        && (httpResponseCode != HttpURLConnection.HTTP_NOT_MODIFIED)) {
+      String errorMessage = getErrorStream(connection);
+
+      // todo(annz) add more specific error handling. NOT_FOUND, etc.
+      return Tasks.forException(
+          new Exception(
+              String.format(
+                  Locale.getDefault(),
+                  "Failed to connect to Firebase ML download server with HTTP status code: %d"
+                      + " and error message: %s",
+                  connection.getResponseCode(),
+                  errorMessage)));
+    } else if (httpResponseCode == HttpURLConnection.HTTP_NOT_MODIFIED) {
+      return Tasks.forResult(null);
+    }
+
+    return Tasks.forResult(readCustomModelResponse(modelName, connection));
+  }
+
+  private CustomModel readCustomModelResponse(
+      @NonNull String modelName, HttpURLConnection connection) throws IOException {
+
+    String encodingKey = connection.getHeaderField(CONTENT_ENCODING_HEADER_KEY);
+    InputStream inputStream = maybeUnGzip(connection.getInputStream(), encodingKey);
+    JsonReader reader = new JsonReader(new InputStreamReader(inputStream, UTF_8));
+    long fileSize = 0L;
+    String downloadUrl = "";
+    long expireTime = 0L;
+
+    String modelHash = maybeUnGzipHeader(connection.getHeaderField(ETAG_HEADER), encodingKey);
+
+    if (modelHash == null || modelHash.isEmpty()) {
+      // todo(annz) replace this...
+      modelHash = connection.getResponseMessage();
+    }
+
+    // JsonReader.peek will sometimes throw AssertionErrors in Android 8.0 and above. See
+    // https://b.corp.google.com/issues/79920590 for details.
```

Comment:
Use b/79920590

(other instances fixed in https://github.com/firebase/firebase-android-sdk/pull/2174)

Reply:
Done.